File: clients/node/squeue.mjs
// Minimal SqueueLite client — Unix Domain Socket, JSON-RPC 2.0.
//
// SqueueLite is *write-only*. Send writes through this client; read straight
// from the SQLite file with a read-only connection (WAL allows readers).
//
// Example:
//   import { Squeue } from "./squeue.mjs";
//
//   const db = await Squeue.connect("/run/app/squeuelite.sock", "agent-node");
//   const r = await db.execute(
//     "INSERT INTO events(agent_id, kind) VALUES (?, ?)",
//     ["agent-node", "started"],
//     { idempotencyKey: "run-7:step-1", runId: "run-7" });
//   // r == { jsonrpc: "2.0", id: 1, result: { status: "committed", commit_seq: 12 } }
//   // Check r.result for success, r.error for failure.
//   db.close();
//
// No third-party dependencies; Node >= 16.

import net from "node:net";

// Wrap binary data for a BLOB column parameter: {"$blob": "<base64>"}.
// `data` may be a Buffer, Uint8Array, or ArrayBuffer.
//   db.execute("INSERT INTO files(data) VALUES (?)", [blob(buf)])
export function blob(data) {
  return { $blob: Buffer.from(data).toString("base64") };
}

export class Squeue {
  #sock;
  #buf = "";
  #waiters = [];
  #counter = 0;

  static connect(socketPath, actorId) {
    return new Promise((resolve, reject) => {
      const sock = net.createConnection(socketPath);
      const self = new Squeue();
      self.#sock = sock;
      self.actorId = actorId;
      sock.setEncoding("utf8");
      sock.once("connect", () => resolve(self));
      sock.once("error", reject);
      sock.on("data", (chunk) => self.#onData(chunk));
      sock.on("close", () => {
        while (self.#waiters.length) {
          self.#waiters.shift().reject(new Error("gateway closed the connection"));
        }
      });
    });
  }

  #onData(chunk) {
    this.#buf += chunk;
    let i;
    while ((i = this.#buf.indexOf("\n")) >= 0) {
      const line = this.#buf.slice(0, i);
      this.#buf = this.#buf.slice(i + 1);
      const w = this.#waiters.shift();
      if (!w) continue;
      try {
        w.resolve(JSON.parse(line));
      } catch (e) {
        w.reject(e);
      }
    }
  }

  #nextId() {
    return ++this.#counter;
  }

  #roundtrip(obj) {
    return new Promise((resolve, reject) => {
      this.#waiters.push({ resolve, reject });
      this.#sock.write(JSON.stringify(obj) + "\n");
    });
  }

  // Run one statement as a single atomic transaction.
  execute(sql, params = [], opts = {}) {
    return this.transaction([[sql, params]], opts);
  }

  // Run several [sql, params] ops as ONE transaction (all-or-nothing).
  transaction(ops, { idempotencyKey, runId } = {}) {
    const rpcParams = {
      actor_id: this.actorId,
      operations: ops.map(([sql, params]) => ({ sql, params })),
    };
    if (runId != null) rpcParams.run_id = runId;
    if (idempotencyKey != null) rpcParams.idempotency_key = idempotencyKey;
    return this.#roundtrip({
      jsonrpc: "2.0",
      id: this.#nextId(),
      method: "execute",
      params: rpcParams,
    });
  }

  stats() {
    return this.#roundtrip({ jsonrpc: "2.0", id: this.#nextId(), method: "stats" });
  }
  health() {
    return this.#roundtrip({ jsonrpc: "2.0", id: this.#nextId(), method: "health" });
  }
  checkpoint() {
    return this.#roundtrip({ jsonrpc: "2.0", id: this.#nextId(), method: "checkpoint" });
  }

  close() {
    this.#sock.end();
  }
}

// Tiny demo:  node squeue.mjs ./squeuelite.sock
if (import.meta.url === `file://${process.argv[1]}`) {
  const path = process.argv[2] ?? "./squeuelite.sock";
  const db = await Squeue.connect(path, "demo-node");
  console.log("health:", await db.health());
  console.log(
    "write :",
    await db.execute("INSERT INTO events(agent_id, kind) VALUES (?, ?)", ["demo-node", "started"], {
      idempotencyKey: "demo:1",
    }),
  );
  console.log("stats :", await db.stats());
  db.close();
}

File: clients/python/squeue.py
"""Minimal SqueueLite client — Unix Domain Socket, JSON-RPC 2.0.

SqueueLite is *write-only*. Send writes through this client; read straight from
the SQLite file with a read-only connection (WAL allows concurrent readers).

Example
-------
    from squeue import Squeue

    with Squeue("/run/app/squeuelite.sock", actor_id="agent-py") as db:
        resp = db.execute(
            "INSERT INTO events(agent_id, kind, payload) VALUES (?, ?, ?)",
            ["agent-py", "started", '{"ok": true}'],
            idempotency_key="run-7:step-1",   # safe to retry
            run_id="run-7",
        )
        # resp == {"jsonrpc": "2.0", "id": 1, "result": {"status": "committed", "commit_seq": 12}}
        # Check resp["result"] for success, resp["error"] for failure.

No third-party dependencies; standard library only.
"""

import base64
import json
import socket


def blob(data: bytes) -> dict:
    """Wrap binary data for a BLOB column parameter: {"$blob": "<base64>"}.

    Example:
        db.execute("INSERT INTO files(data) VALUES (?)", [blob(b"\\x00\\x01\\x02")])
    """
    return {"$blob": base64.b64encode(data).decode("ascii")}


class SqueueError(RuntimeError):
    """Raised when the gateway closes the connection unexpectedly."""


class Squeue:
    def __init__(self, socket_path: str, actor_id: str):
        self.actor_id = actor_id
        self._counter = 0
        self._sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        self._sock.connect(socket_path)
        # Buffered, newline-aware IO in both directions.
        self._io = self._sock.makefile("rwb")

    def _next_id(self) -> int:
        self._counter += 1
        return self._counter

    # -- write API ---------------------------------------------------------

    def execute(self, sql, params=None, *, idempotency_key=None, run_id=None):
        """Run one statement as a single atomic transaction."""
        return self.transaction(
            [(sql, params or [])],
            idempotency_key=idempotency_key,
            run_id=run_id,
        )

    def transaction(self, ops, *, idempotency_key=None, run_id=None):
        """Run several (sql, params) ops as ONE transaction (all-or-nothing)."""
        rpc_params = {
            "actor_id": self.actor_id,
            "operations": [{"sql": s, "params": list(p)} for (s, p) in ops],
        }
        if run_id is not None:
            rpc_params["run_id"] = run_id
        if idempotency_key is not None:
            rpc_params["idempotency_key"] = idempotency_key
        req = {
            "jsonrpc": "2.0",
            "id": self._next_id(),
            "method": "execute",
            "params": rpc_params,
        }
        return self._roundtrip(json.dumps(req))

    # -- admin API ---------------------------------------------------------

    def stats(self):
        req = {"jsonrpc": "2.0", "id": self._next_id(), "method": "stats"}
        return self._roundtrip(json.dumps(req))

    def health(self):
        req = {"jsonrpc": "2.0", "id": self._next_id(), "method": "health"}
        return self._roundtrip(json.dumps(req))

    def checkpoint(self):
        req = {"jsonrpc": "2.0", "id": self._next_id(), "method": "checkpoint"}
        return self._roundtrip(json.dumps(req))

    # -- internals ---------------------------------------------------------

    def _roundtrip(self, line: str):
        self._io.write(line.encode() + b"\n")
        self._io.flush()
        reply = self._io.readline()
        if not reply:
            raise SqueueError("gateway closed the connection")
        return json.loads(reply)

    def close(self):
        try:
            self._io.close()
        finally:
            self._sock.close()

    def __enter__(self):
        return self

    def __exit__(self, *_exc):
        self.close()


if __name__ == "__main__":
    import sys

    sock_path = sys.argv[1] if len(sys.argv) > 1 else "./squeuelite.sock"
    with Squeue(sock_path, actor_id="demo-py") as db:
        resp = db.health()
        print("health:", resp)
        resp = db.execute(
            "INSERT INTO events(agent_id, kind) VALUES (?, ?)",
            ["demo-py", "started"],
            idempotency_key="demo:1",
        )
        print("write :", resp)
        print("stats :", db.stats())
